In [121]:
import pandas as pd
import numpy as np
import pickle

In [122]:
f = open("./final_data.p", "rb")
data = pickle.load(f)

In [123]:
data.head()

fips_code              county  total_2008  dem_2008  gop_2008  oth_2008  \
0      26041        Delta County     19064.0    9974.0    8763.0     327.0   
1      48295     Lipscomb County      1256.0     155.0    1093.0       8.0   
2       1127       Walker County     28652.0    7420.0   20722.0     510.0   
3      48389       Reeves County      3077.0    1606.0    1445.0      26.0   
4      56017  Hot Springs County      2546.0     619.0    1834.0      93.0   

   total_2012  dem_2012  gop_2012  oth_2012          ...           \
0     18043.0    8330.0    9533.0     180.0          ...            
1      1168.0     119.0    1044.0       5.0          ...            
2     28497.0    6551.0   21633.0     313.0          ...            
3      2867.0    1649.0    1185.0      33.0          ...            
4      2495.0     523.0    1894.0      78.0          ...            

   Family Income 35-40K  Family Income 40-45K  Family Income 45-50K  \
0                 453.0                 572.0                 528.0   
1                  20.0                  29.0                  33.0   
2                1041.0                 919.0                 987.0   
3                  96.0                 119.0                  75.0   
4                  50.0                  69.0                  84.0   

   Family Income 50-60K  Family Income 60-75K  Family Income 75-100K  \
0                1217.0                1261.0                 1697.0   
1                  79.0                 123.0                  116.0   
2                1570.0                2209.0                 1819.0   
3                 213.0                 320.0                  350.0   
4                 122.0                 234.0                  112.0   

   Family Income 100-125K  Family Income 125-150K  Family Income 150-200K  \
0                   623.0                   401.0                   234.0   
1                    91.0                    54.0                    56.0   
2                  1210.0                   745.0                   437.0   
3                   188.0                    69.0                    64.0   
4                   148.0                    89.0                    36.0   

   Family Income > 200K  
0                  96.0  
1                  46.0  
2                 339.0  
3                  79.0  
4                  58.0  

[5 rows x 134 columns]

In [126]:
X = data.ix[:, 14:]

In [130]:
X.head()

# of Housing Units  # White  # White Male  # White Female  # Black  \
0              3040.0  34754.0       17231.0         17523.0    121.0   
1               478.0   2980.0        1577.0          1403.0     17.0   
2              2700.0  60224.0       29335.0         30889.0   3764.0   
3               556.0  11899.0        6966.0          4933.0    750.0   
4               326.0   4661.0        2298.0          2363.0     12.0   

   # Black Male  # Black Female  # White Not Hispanic  \
0          60.0            61.0               34581.0   
1           8.0             9.0                2361.0   
2        1781.0          1983.0               59498.0   
3         578.0           172.0                2677.0   
4           1.0            11.0                4535.0   

   # White Not Hispanis Male  # White Not Hispanis Female  \
0                    17137.0                      17444.0   
1                     1194.0                       1167.0   
2                    28884.0                      30614.0   
3                     1795.0                        882.0   
4                     2247.0                       2288.0   

           ...           Family Income 35-40K  Family Income 40-45K  \
0          ...                          453.0                 572.0   
1          ...                           20.0                  29.0   
2          ...                         1041.0                 919.0   
3          ...                           96.0                 119.0   
4          ...                           50.0                  69.0   

   Family Income 45-50K  Family Income 50-60K  Family Income 60-75K  \
0                 528.0                1217.0                1261.0   
1                  33.0                  79.0                 123.0   
2                 987.0                1570.0                2209.0   
3                  75.0                 213.0                 320.0   
4                  84.0                 122.0                 234.0   

   Family Income 75-100K  Family Income 100-125K  Family Income 125-150K  \
0                 1697.0                   623.0                   401.0   
1                  116.0                    91.0                    54.0   
2                 1819.0                  1210.0                   745.0   
3                  350.0                   188.0                    69.0   
4                  112.0                   148.0                    89.0   

   Family Income 150-200K  Family Income > 200K  
0                   234.0                  96.0  
1                    56.0                  46.0  
2                   437.0                 339.0  
3                    64.0                  79.0  
4                    36.0                  58.0  

[5 rows x 120 columns]

In [135]:
del X["Population"]

In [147]:
X = X.apply(lambda x: x / data["Population"])

In [148]:
voter_turn_out = data["total_2016"] - data["total_2012"]

In [149]:
from sklearn.linear_model import LinearRegression

In [150]:
X.shape

(2628, 119)

In [158]:
linear = LinearRegression()

In [159]:
linear.fit(X, voter_turn_out)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [160]:
for name, val in sorted(zip(X.columns, linear.coef_), key=lambda x: x[1], reverse=True):
    print((name, val))

('Median Age Male', 3527386.0299658948)
('Median Age Female', 1490393.5966894927)
('Rail Road', 881437.66146147694)
('Family Income > 200K', 304990.08713594591)
('Moved From Abroad White', 267502.00375204743)
('HS or equal', 265866.51162106113)
('Total Moving White', 258568.70454609708)
('Other State-Bachelor', 249590.51199860801)
('Graduate', 245747.16516818199)
('Family Income 125-150K', 228419.92306020926)
('Family Income 150-200K', 225019.83121520071)
('No Vehicle Owned', 222998.41806845431)
('In State-Less than HS', 213484.41168846484)
('Family Income 100-125K', 209364.59985931509)
('Other State-Less than HS', 172594.16588157462)
('In State-Bachelor', 163692.91933648489)
('Family Income 75-100K', 136615.99178435892)
('Public Transportation', 104600.14325030538)
('Income < 10K', 92003.62800139327)
('Family Income 60-75K', 84445.569386091054)
('Other State-Some College', 81802.963755326578)
('In State-Some College', 81211.617145977856)
('# Male', 70939.649665328761)
('Total Reporing

In [161]:
from sklearn.linear_model import LassoCV

In [162]:
lasso = LassoCV()

In [164]:
lasso.fit(X, voter_turn_out)

/opt/conda/lib/python3.5/site-packages/sklearn/linear_model/coordinate_descent.py:479: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)


LassoCV(alphas=None, copy_X=True, cv=None, eps=0.001, fit_intercept=True,
    max_iter=1000, n_alphas=100, n_jobs=1, normalize=False, positive=False,
    precompute='auto', random_state=None, selection='cyclic', tol=0.0001,
    verbose=False)

In [169]:
keep_vars = []

for name, val in zip(X.columns, lasso.coef_):    
    if val != 0:
        keep_vars.append(name)

In [170]:
linear = LinearRegression()

In [171]:
linear.fit(X[keep_vars], y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [172]:
for name, val in sorted(zip(X.columns, linear.coef_), key=lambda x: x[1], reverse=True):
    if val != 0:
        print((name, val))

('Other State-Less than HS', 153835.48684836502)
('# Hispanic Male', 104398.6050783024)
('Total Reporting Income', 97516.669941037922)
('Other State-Bachelor', 78723.800301144423)
('# Female', 76520.067351112346)
('In State-Some College', 74757.104745037039)
('Other State-Graduate', 67699.326989439491)
('Other State-Some College', 57047.995538552721)
('# Hispanic Female', 48568.23806044356)
('Median Age Male', 43713.529341899753)
('Other State-HS or equal', 42556.725064011342)
('In State-HS or equal', 34664.628344767858)
('Divorced', 34210.787776993282)
('# of Housing Units', 32481.828833846601)
('Graduate', 27934.96635801471)
('Median Age Female', 14719.334554924908)
('# White Male', 12019.596271139255)
('In State-Graduate', 8485.3685540876395)
('Median Age', 7247.5357488882382)
('# Black', 5560.9469048359624)
('# Black Male', 4548.8297617578401)
('Some College-AA', 3451.4785371167895)
('Bachelor', 1396.1108441553417)
('Citizen by Naturalization', 414.53045064552731)
('Never Married',

How some of these can be interpretted.

('# Female', 76520.067351112346)

After adjusting for other predictors, each 1% increase in the number of females in a county is associated with an increase in the mean voter turn out from 2012 to 2016 by 76520 voters.

In [91]:
diff_voting_2012 = data["dem_2012"] - data["gop_2012"]
diff_voting_2016 = data["dem_2016"] - data["gop_2016"]

y = diff_voting_2016 - diff_voting_2012

In [99]:
ratio_2012 = data["dem_2012"] / data["total_2012"] - 0.5
ratio_2016 = data["dem_2016"] / data["total_2016"] - 0.5

ratio_change = diff_voting_2016 - diff_voting_2012